crates/typst/src/export/svg.rs

```diff
@@ -0,0 +1,452 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Write},
+    io::Read,
+};
+
+use base64::Engine;
+use ttf_parser::{GlyphId, OutlineBuilder};
+use usvg::{NodeExt, TreeParsing};
+
+use crate::{
+    doc::{Frame, Glyph, TextItem},
+    geom::{Abs, Axes, Shape, Transform},
+    util::hash128,
+};
+use crate::{geom::Paint::Solid, image::Image};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct RenderHash(u128);
+
+impl From<u128> for RenderHash {
+    fn from(value: u128) -> Self {
+        Self(value)
+    }
+}
+
+impl Display for RenderHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        base64::engine::general_purpose::STANDARD
+            .encode(self.0.to_le_bytes())
+            .fmt(f)
+    }
+}
+
+/// Export a document into a SVG file.
+pub fn svg(page: &Frame) -> String {
+    let mut renderer = SVGRenderer::default();
+    let page_string = renderer.render_page(page, Transform::identity());
+    renderer.append_page(page_string);
+    renderer.finalize(page.size())
+}
+
+#[derive(Debug, Clone, Default)]
+struct SVGRenderer {
+    body: String,
+    glyphs: HashMap<RenderHash, String>,
+    clip_paths: HashMap<RenderHash, String>,
+}
+
+impl SVGRenderer {
+    fn header(&self, size: Axes<Abs>) -> String {
+        let mut res = format!(
+            r#"<svg class="typst-doc" viewBox="0 0 {0} {1}" width="{0}" height="{1}"
```

Comment:
Would it perhaps make sense to use the `xmlwriter` crate here? This would make this a little bit more high-level to write and also give the option to easily format with indent etc. I would even go as far as to have readable indent by default as the size difference is pretty small and the resulting files are far easier to digest. (Typst's PDF actually also use readable indent.)

Reply:
`xmlwrite` makes sense to me. But it seems that this crate wasn't very popular(only 18 stars). Perhaps we can use something similar, but more popular and stable.

Reply:
Have to say that `xmlwrite`'s api looks nice to me. And the codebase is quite small. I'll give it a try.

Reply:
I like RazrFalcon's crates. They are good stuff.

Reply:
`xmlwrite` means that you must hold a state (the writing string stream)  in the `SvgRenderer`, hence fully prevent incremental things. But it is sufficiently enough to only memorize glyph rendering to my observation.

Reply:
It seems that `xmlwrite` aims to create an XML document linearly(by linearly I mean it is impossible to insert element in the middle.), but we need to first build the document, collect used glyphs and put glys in the `<defs>` tag. Therefore, string concatenation may still be used, but not as frequently as the current implementation. Incidentally, I just realized that if possible, we could put glyph defs at the end of the document to prevent string concatenation.